### ISSUE TITLE
Metric spaces

### ISSUE THREAD

**Issue by brendanzab** [sebcrozet/alga#20]
cgmath currently has a [`MetricSpace`](https://docs.rs/cgmath/0.15.0/cgmath/trait.MetricSpace.html) trait that is defined for vectors, points, and quaternions:

```rust
pub trait MetricSpace: Sized {
    type Metric: BaseFloat;
    fn distance2(self, other: Self) -> Self::Metric;

    fn distance(self, other: Self) -> Self::Metric { ... }
}
```

Could we consider rejiggering the `alga::linear` trait hierarchy to add something like this?

**Comment by sebcrozet** [comment]
Yes, adding metric spaces would be useful. Though I am still not sure what the best definition would be because of the link between dot product (already available on the `FiniteDimInnerSpace`) and the metric it induces. But any suggestion is welcome!
